### PR TITLE
Fix address assignment and binding

### PR DESCRIPTION
**Pull request by astaff**
# Overview

With this fix user will be able to:
- discover USB connected robot faster
- turn a USB connected robot on without having to unplug/plug USB cable
- after Resin container update, clean local packages/opentrons to make container's API take precedence

Development flow:
- On the robot, use `calibrate` to call calibration tool
- Push updates using `make -C api push` without having to restart a robot if it was unplugged
- `make -C api term` to connect over ssh immediately after plugging a USB robot, ignores warning about key mismatch since different robots have different keys, but same names when connected over USB

# Changes

- Switch from `nmcli` to lower level `iproute2` to set up persistent IPv6 address on `eth0`
- Ignore Duplicate Address Detection (DAD) which caused robot to pick a new random address in some cases, when power cycled while being plugged into USB
- Aggressively clean up `nmcli`'s eth0 and unassigned connections to avoid interference
- Give UUID to each new container, and check/update persisted /data/id to detect if it's a first run of a new container and persisted data needs to be updated

Tested on a robot 👍

### CODE REVIEW

File: compute/scripts/setup.sh [hunk 1]
```diff
@@ -10,17 +10,34 @@
...
+# Clean up opentrons package dir if it's a first start of a new container
+touch /data/id
+if [ $(cat /data/id) != $CONTAINER_ID ] ; then 
+  echo 'First start of a new container. Deleting local Opentrons API installation'
+  rm -rf /data/packages/usr/local/lib/python3.6/site-packages/opentrons*
+  echo $CONTAINER_ID > /data/id
+fi
+
+# Set static address so we can find the device from host computer over
+# USB without using Bojnjour or any kind of service discovery, making
```
Comment (btmorr) [inline compute/scripts/setup.sh]: "over ethernet". I know that we are calling this USB internally, and why, but to the developer this looks very strange to refer to the eth0 connection as "over USB"

File: compute/scripts/setup.sh [hunk 2]
```diff
@@ -10,17 +10,34 @@
...
-  con-name 'local-ethernet' \
-  autoconnect yes \
-  type ethernet \
-  ifname eth0 \
-  ipv6.method manual \
-  ipv6.address $ETHERNET_STATIC_IP/$ETHERNET_NETWORK_PREFIX_LENGTH \
-  ipv4.method disabled
+# Disable duplicate address detection (DAD)
+# Leaving DAD enabled sometimes results in static address being reset
+# to a random value after robot restart with USB plugged in,
```
Comment (mcous) [inline compute/scripts/setup.sh]: Same comment as @btmorr lower in the file

File: Dockerfile [hunk 3]
```diff
@@ -85,6 +85,10 @@ RUN sed -i "s/{ETHERNET_NETWORK_PREFIX}/$ETHERNET_NETWORK_PREFIX/g" /etc/radvd.c
 ENV PIP_ROOT /data/packages
 RUN echo "export PIP_ROOT=$PIP_ROOT" >> /etc/profile
 
+# Generate the id that we will later check to see if that's the
+# new container and that local Opentrons API package should be deleted
+ENV CONTAINER_ID=$(uuidgen)
```
Comment (mcous) [inline Dockerfile]: Was it not possible to use the regular docker container ID?

File: compute/scripts/setup.sh [hunk 4]
```diff
@@ -10,17 +10,34 @@
...
 
+# Cleanup any connections. This will leave only wlan0
+nmcli --terse --fields uuid,device connection show | sed -rn 's/(.*):(--)/\1/p' | xargs nmcli connection del
+nmcli --terse --fields uuid,device connection show | sed -rn 's/(.*):(eth0)/\1/p' | xargs nmcli connection del
+
+# Clean up opentrons package dir if it's a first start of a new container
+touch /data/id
+if [ $(cat /data/id) != $CONTAINER_ID ] ; then 
+  echo 'First start of a new container. Deleting local Opentrons API installation'
+  rm -rf /data/packages/usr/local/lib/python3.6/site-packages/opentrons*
```
Comment (mcous) [inline compute/scripts/setup.sh]: First off, why is it `python3.6`? I thought our environment was 3.5.3. Second off, this seems seriously hacky and indicative that we are not setting up this container properly. The whole point of a Docker container is to wrap an application; are we storing the application outside of the container or something?

**Comment by btmorr** [comment]
Tried to connect after reboot. Got this in my terminal:

attemtreconafterreboot

And this was the log in resin:

attemtreconafterreboot2

File: Dockerfile (hunk 3 above)
Comment (astaff) [reply Dockerfile]: Not that i am aware of. Would appreciate your help in finding the answer.

File: compute/scripts/setup.sh (hunk 4 above)
Comment (astaff) [reply compute/scripts/setup.sh]: We are allowing to overlay container's API with the package pushed over http endpoint listening on Ethernet port. This also allows the user to install packages that can be used by the protocol. It's been an ongoing request for app 2.5.2 where python environment is sandbox within pyinstaller.

I don't think canonical docker architecture was meant to be used on constrained devices and I am convinced we should be making exceptions when building containers for devices. While resin is great it doesn't allow to side-load image, it only allows for image to be built and executed on the device.

Being able to silde-load images and add packages in an important use case and this seems to be able to address it.

Python 3.5.3 was a result of pyinstaller dependency which didn't support 3.5.3. We should bump it up to 3.6.

File: Dockerfile (hunk 3 above)
Comment (mcous) [reply Dockerfile]: If you didn't find this same SO answer, looks like the docker container ID is present in `/proc/self/cgroup`. Tested on mike-bot PI and verified that what's there definitely looks like a container ID. Source: https://stackoverflow.com/questions/20995351/docker-how-to-get-container-information-from-within-the-container#25729598

Comment (astaff) [reply Dockerfile]: Great! Not familiar with this approach. Please go ahead and make the change if you feel it is needed.

File: compute/scripts/setup.sh (hunk 2 above)
Comment (mcous) [reply compute/scripts/setup.sh]: Bump

**Pull request merged**